File: Legacy/Python/IsTarget.py
import os
import cv2
import sys
from adb_helper import ADBHelper
from vision_engine import VisionEngine

# Tọa độ vùng quét tài nguyên ở độ phân giải 1600x900px
COORDS = {
    "Gold": (55, 117, 251, 161),
    "Elixir": (60, 167, 261, 208),
    "Dark Elixir": (73, 214, 183, 248)
}

# Lề (margin) cắt bù trừ cho từng loại tài nguyên
PADDING = {
    "Gold": {"l": 60, "r": 15, "t": 5, "b": 5},
    "Elixir": {"l": 15, "r": 15, "t": 5, "b": 5},
    "Dark Elixir": {"l": 15, "r": 15, "t": 5, "b": 5}
}

def extract_resources(adb: ADBHelper = None, vision: VisionEngine = None) -> tuple[int, int, int]:
    """
    Trích xuất chỉ số tài nguyên hiện tại từ giả lập.
    Được thiết kế siêu nhẹ, loại bỏ PyTorch/EasyOCR bằng thuật toán khớp mẫu chữ số.
    """
    if adb is None:
        adb = ADBHelper()
    if vision is None:
        vision = VisionEngine()

    print("[SCOUT] Đang chụp ảnh màn hình để phân tích tài nguyên...")
    screenshot = adb.take_screenshot()
    if screenshot is None:
        print("[SCOUT ERROR] Không thể chụp ảnh màn hình từ giả lập.")
        return 0, 0, 0

    h_img, w_img, _ = screenshot.shape
    results = {}

    for label, (x1, y1, x2, y2) in COORDS.items():
        p = PADDING[label]
        
        # Tính toán tọa độ cắt bù trừ an toàn
        x1p = max(0, x1 - p["l"])
        y1p = max(0, y1 - p["t"])
        x2p = min(w_img, x2 + p["r"])
        y2p = min(h_img, y2 + p["b"])
        
        # Cắt ROI tài nguyên và nhận dạng chữ số qua VisionEngine
        roi = (x1p, y1p, x2p - x1p, y2p - y1p)
        val = vision.extract_numerical_metrics(screenshot, roi)
        results[label] = val

    gold = results.get("Gold", 0)
    elixir = results.get("Elixir", 0)
    dark_elixir = results.get("Dark Elixir", 0)

    print(f"[SCOUT] Kết quả quét -> Vàng: {gold:,} | Dầu hồng: {elixir:,} | Dầu đen: {dark_elixir:,}")
    return gold, elixir, dark_elixir

if __name__ == "__main__":
    # Điểm chạy độc lập để lập trình viên kiểm thử nhanh
    adb_instance = ADBHelper()
    vision_instance = VisionEngine()
    extract_resources(adb_instance, vision_instance)

File: src/Simplimixi/Core/IsTarget.cs
using System;
using System.Collections.Generic;
using OpenCvSharp;

namespace CvAut
{
    /// <summary>
    /// Lớp tiện ích cung cấp các toạ độ và chức năng để quét, phân tích tài nguyên (Vàng, Dầu hồng, Dầu đen)
    /// từ các màn hình của Clash of Clans. Được thiết kế tối ưu cho độ phân giải giả lập chuẩn 1600x900px.
    /// </summary>
    public static class IsTarget
    {
        // Tọa độ vùng quét tài nguyên đối thủ khi đang đi tìm trận (Scout Screen)
        // Dựa trên chuẩn giao diện độ phân giải 1600x900px
        private static readonly Dictionary<string, Rect> Coords = new()
        {
            { "Gold", new Rect(55, 117, 196, 44) },        // Vàng: x=55, y=117, rộng=196, cao=44 (Vùng hiển thị số lượng vàng cướp được)
            { "Elixir", new Rect(60, 167, 201, 41) },      // Dầu hồng: x=60, y=167, rộng=201, cao=41
            { "Dark Elixir", new Rect(73, 214, 110, 34) }  // Dầu đen: x=73, y=214, rộng=110, cao=34
        };

        // Lề cắt rộng bù trừ margin để đảm bảo lấy trọn chữ số, tránh bị mất viền của ký tự đầu hoặc cuối
        private static readonly Dictionary<string, Padding> Margins = new()
        {
            { "Gold", new Padding { L = 60, R = 15, T = 5, B = 5 } }, // Thêm lề trái rộng do số vàng có thể dài
            { "Elixir", new Padding { L = 15, R = 15, T = 5, B = 5 } },
            { "Dark Elixir", new Padding { L = 15, R = 15, T = 5, B = 5 } }
        };

        /// <summary>
        /// Cấu trúc lưu trữ khoảng lề bù trừ xung quanh vùng quét.
        /// </summary>
        private struct Padding
        {
            public int L { get; set; } // Left (Lề Trái)
            public int R { get; set; } // Right (Lề Phải)
            public int T { get; set; } // Top (Lề Trên)
            public int B { get; set; } // Bottom (Lề Dưới)
        }

        /// <summary>
        /// Thực hiện chụp ảnh màn hình giả lập và trích xuất các chỉ số tài nguyên hiện tại của nhà đối thủ (khi tìm trận).
        /// </summary>
        /// <param name="adb">Đối tượng ADBHelper để thực hiện giao tiếp với thiết bị.</param>
        /// <param name="vision">Đối tượng VisionEngine chứa thuật toán OCR.</param>
        /// <returns>Bộ ba số nguyên biểu thị lượng (Vàng, Dầu hồng, Dầu đen) nhận diện được.</returns>
        public static (int Gold, int Elixir, int DarkElixir) ExtractResources(ADBHelper adb, VisionEngine vision)
        {
            Console.WriteLine("[SCOUT] Reading target resources...");
            using Mat? screenshot = adb.TakeScreenshot();
            if (screenshot == null || screenshot.Empty())
            {
                Console.WriteLine("[SCOUT ERROR] Screenshot unavailable.");
                return (0, 0, 0);
            }

            int hImg = screenshot.Height;
            int wImg = screenshot.Width;
            var results = new Dictionary<string, int>();

            foreach (var kvp in Coords)
            {
                string label = kvp.Key;
                Rect r = kvp.Value;
                Padding p = Margins[label];

                // Tính toán tọa độ cắt bù trừ an toàn chống lỗi tràn biên (Out of bounds)
                int x1p = Math.Max(0, r.X - p.L);
                int y1p = Math.Max(0, r.Y - p.T);
                int x2p = Math.Min(wImg, r.X + r.Width + p.R);
                int y2p = Math.Min(hImg, r.Y + r.Height + p.B);

                Rect roi = new Rect(x1p, y1p, x2p - x1p, y2p - y1p);

                // Gọi bộ phân tích OCR của VisionEngine để đọc số trong vùng chọn (ROI)
                int val = vision.ExtractNumericalMetrics(screenshot, roi);
                results[label] = val;
            }

            int gold = results.GetValueOrDefault("Gold", 0);
            int elixir = results.GetValueOrDefault("Elixir", 0);
            int darkElixir = results.GetValueOrDefault("Dark Elixir", 0);

            Console.WriteLine($"[SCOUT] Resources: Gold={gold:N0}, Elixir={elixir:N0}, Dark={darkElixir:N0}");
            return (gold, elixir, darkElixir);
        }

        // Tọa độ vùng quét tài nguyên Làng chính (Home Base) chuẩn độ phân giải 1600x900px ở góc trên cùng bên phải màn hình
        private static readonly Dictionary<string, Rect> HomeCoords = new()
        {
            { "Gold", new Rect(1310, 30, 200, 36) },       // Vùng chứa thông số Vàng của Làng chính
            { "Elixir", new Rect(1310, 115, 200, 36) },     // Vùng chứa thông số Dầu hồng của Làng chính
            { "Dark Elixir", new Rect(1310, 200, 200, 32) } // Vùng chứa thông số Dầu đen của Làng chính
        };

        /// <summary>
        /// Thực hiện chụp ảnh màn hình giả lập và trích xuất chỉ số tài nguyên hiện có ở Làng chính (Home Base).
        /// Dùng để phục vụ các quyết định nâng cấp tường hoặc công trình.
        /// </summary>
        /// <param name="adb">Đối tượng ADBHelper.</param>
        /// <param name="vision">Đối tượng VisionEngine.</param>
        /// <returns>Bộ ba số nguyên biểu thị lượng tài nguyên làng chính.</returns>
        public static (int Gold, int Elixir, int DarkElixir) ExtractHomeResources(ADBHelper adb, VisionEngine vision)
        {
            Console.WriteLine("[SCOUT] Reading home resources...");
            using Mat? screenshot = adb.TakeScreenshot();
            if (screenshot == null || screenshot.Empty())
            {
                Console.WriteLine("[SCOUT ERROR] Screenshot unavailable.");
                return (0, 0, 0);
            }

            var results = new Dictionary<string, int>();

            foreach (var kvp in HomeCoords)
            {
                string label = kvp.Key;
                Rect roi = kvp.Value;

                // Sử dụng RGB thresholding (useRgbThresh: true) để loại bỏ nền vàng/hồng sáng màu của thanh chứa tài nguyên Làng chính,
                // chỉ giữ lại phần màu chữ số trắng/đen để chạy OCR chính xác hơn.
                int val = vision.ExtractNumericalMetrics(screenshot, roi, isOffline: false, useRgbThresh: true);
                results[label] = val;
            }

            int gold = results.GetValueOrDefault("Gold", 0);
            int elixir = results.GetValueOrDefault("Elixir", 0);
            int darkElixir = results.GetValueOrDefault("Dark Elixir", 0);

            Console.WriteLine($"[SCOUT] Home resources: Gold={gold:N0}, Elixir={elixir:N0}, Dark={darkElixir:N0}");
            return (gold, elixir, darkElixir);
        }
    }
}
